File: backend/app/models/journey.py
from datetime import date, datetime

from pydantic import Field

from .common import CamelModel


class CourseStopResponse(CamelModel):
    order: int
    time_slot: str = Field(alias="timeSlot")
    place_id: int = Field(alias="placeId")
    quest_id: int | None = Field(default=None, alias="questId")
    name: str | None = None
    category: str | None = None
    address: str | None = None
    latitude: float
    longitude: float
    img: str | None = None
    icon: str
    menu: str | None = None
    # '연중무휴' / '월,목' / None(모름)
    rest_date: str | None = Field(default=None, alias="restDate")
    # 화면 상세 카드에 보여 주는 실용 정보
    operating_hours: str | None = Field(default=None, alias="operatingHours")
    parking: str | None = None
    hours_unknown: bool = Field(alias="hoursUnknown")
    distance_km: float = Field(alias="distanceKm")
    # 사용자가 방문 완료로 저장한 퀘스트인지
    completed: bool = False
    # 도슨트로 읽어 줄 설명이 있는지
    docent: bool = False


class CourseResponse(CamelModel):
    course_id: int | None = Field(default=None, alias="courseId")
    persona_key: str = Field(alias="personaKey")
    persona_name: str = Field(alias="personaName")
    visit_date: date = Field(alias="visitDate")
    total_km: float = Field(alias="totalKm")
    stops: list[CourseStopResponse]
    # 후보가 없어 채우지 못한 시간대
    skipped_slots: list[str] = Field(default_factory=list, alias="skippedSlots")


class CharacterResponse(CamelModel):
    key: str
    name: str


class DocentResponse(CamelModel):
    place_id: int = Field(alias="placeId")
    name: str | None = None
    # PLACE.TEXT 원문 전체. 화면에 같이 보여 주고, 음성도 이 글을 읽는다.
    text: str
    source: str = "한국관광공사"


class QuestCompletionResponse(CamelModel):
    quest_id: int = Field(alias="questId")
    name: str | None = None
    completed: bool
    completed_at: datetime | None = Field(default=None, alias="completedAt")

File: frontend/lib/api/journey.js
import { ApiError } from "./spots";
import { createTranslator } from "../i18n";

const API_BASE_URL =
  process.env.NEXT_PUBLIC_API_BASE_URL || "http://localhost:8001";
// 로그인이 붙기 전까지는 테스트 사용자(USERS.NO = 1)로 저장한다.
// 로그인이 생기면 이 값만 로그인한 사용자 번호로 바꾸면 된다.
const LOCAL_USER_NO = 1;

// 코스 API 응답을 지도 컴포넌트가 쓰는 장소 모양으로 바꾼다.
// (id / name / description / distance / latitude / longitude / icon / docent)
function toPlace(stop, t) {
  const details = [stop.timeSlot, stop.category].filter(Boolean).join(" · ");
  const notice = stop.hoursUnknown ? t("map.closedNotice") : "";
  return {
    id: `place-${stop.placeId}`,
    placeId: stop.placeId,
    // 방문 완료 저장에 쓰는 퀘스트 번호 (QUEST.IDX)
    questId: stop.questId,
    completed: Boolean(stop.completed),
    name: stop.name,
    description: `${details}${notice}`,
    // 상세 카드에 줄줄이 보여 주는 실용 정보
    menu: stop.menu,
    operatingHours: stop.operatingHours,
    parking: stop.parking,
    restDate: stop.restDate,
    distance: stop.order === 1 ? t("map.start") : `${stop.distanceKm}km`,
    latitude: stop.latitude,
    longitude: stop.longitude,
    icon: stop.icon,
    // 서버에 읽어 줄 설명(PLACE.TEXT)이 있으면 도슨트 버튼이 켜진다.
    docent: Boolean(stop.docent),
    timeSlot: stop.timeSlot,
    order: stop.order,
    img: stop.img,
    address: stop.address,
  };
}

async function requestCourse({ persona, date, language = "ko", method, signal, t = createTranslator(language) }) {
  const parameters = new URLSearchParams({ persona });
  if (date) parameters.set("date", date);
  parameters.set("lang", language);

  const path = method === "POST" ? "/api/v1/journey/course/refresh" : "/api/v1/journey/course";
  let response;
  try {
    response = await fetch(`${API_BASE_URL}${path}?${parameters}`, {
      method,
      headers: { "X-User-No": String(LOCAL_USER_NO) },
      signal,
    });
  } catch (error) {
    if (error.name === "AbortError") throw error;
    throw new ApiError("서버에 연결하지 못했습니다.", "NETWORK_ERROR", null);
  }

  const payload = await response.json().catch(() => ({}));
  if (!response.ok) {
    throw new ApiError(
      payload.error?.message || "코스를 불러오지 못했습니다.",
      payload.error?.code || "UNKNOWN_ERROR",
      response.status,
    );
  }

  return { ...payload, places: (payload.stops || []).map((stop) => toPlace(stop, t)) };
}

// 저장된 코스를 준다. 없으면 서버가 이때 한 번 뽑아 저장한다.
export async function fetchCourse({ persona = "king", date, language, signal, t } = {}) {
  return requestCourse({ persona, date, language, method: "GET", signal, t });
}

// 코스를 새로 뽑아 저장한다. 이전 코스는 비활성으로 내려간다.
export async function refreshCourse({ persona = "king", date, language, signal, t } = {}) {
  return requestCourse({ persona, date, language, method: "POST", signal, t });
}

// 마지막에 고른 역할. 앱을 다시 열었을 때 그 역할로 돌아가기 위해 쓴다.
// 아직 고른 적이 없으면 null 을 준다. (처음 들어온 사용자)
export async function fetchSelectedRole({ signal } = {}) {
  let response;
  try {
    response = await fetch(`${API_BASE_URL}/api/v1/journey/characters/selected`, {
      headers: { "X-User-No": String(LOCAL_USER_NO) },
      signal,
    });
  } catch (error) {
    if (error.name === "AbortError") throw error;
    return null;
  }
  if (!response.ok) return null;
  return response.json().catch(() => null);
}

// 방문 완료를 서버에 저장한다. 저장해 두면 다시 들어와도 완료 상태가 남는다.
export async function completeQuest({ questId, signal } = {}) {
  if (!questId) throw new ApiError("퀘스트 번호가 없습니다.", "QUEST_ID_MISSING", null);

  let response;
  try {
    response = await fetch(`${API_BASE_URL}/api/v1/journey/quests/${questId}/complete`, {
      method: "POST",
      headers: { "X-User-No": String(LOCAL_USER_NO) },
      signal,
    });
  } catch (error) {
    if (error.name === "AbortError") throw error;
    throw new ApiError("서버에 연결하지 못했습니다.", "NETWORK_ERROR", null);
  }

  const payload = await response.json().catch(() => ({}));
  if (!response.ok) {
    throw new ApiError(
      payload.error?.message || "방문 완료를 저장하지 못했습니다.",
      payload.error?.code || "UNKNOWN_ERROR",
      response.status,
    );
  }
  return payload;
}
